libgadget/lenstools.c

```diff
@@ -0,0 +1,336 @@
+#include <mpi.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <fftw3.h> 
+#include <string.h>
+
+#ifdef USE_CFITSIO
+#include "fitsio.h"
+#endif
+
+#include "lenstools.h"
+#include "partmanager.h"
+#include "cosmology.h"
+#include "physconst.h"
+#include "utils.h"
+
+void linspace(double start, double stop, int num, double *result) {
+    double step = (stop - start) / (num - 1);
+    for (int i = 0; i < num; i++) {
+        result[i] = start + i * step;
+    }
+}
+
+// Function to allocate a 1D array to be used as a 3D array, and initialize elements to zero
+double *allocate_3d_array_as_1d(int Nx, int Ny, int Nz) {
+    double *array = mymalloc("3d", Nx * Ny * Nz * sizeof(double));
+    // Initialize elements with memset
+    memset(array, 0, Nx * Ny * Nz * sizeof(double));
+    return array;
+}
+
+// Function to allocate a 1D array to be used as a 2D array, and initialize elements to zero
+double *allocate_2d_array_as_1d(int Nx, int Ny) {
+    double *array = mymalloc("2d", Nx * Ny * sizeof(double));
+    // Initialize elements with memset
+    memset(array, 0, Nx * Ny * sizeof(double));
+
+    return array;
+}
+
+typedef struct {
+    int nx, ny, nz;
+} GridDimensions;
+
+// Function to determine the bin index for a given value
+int find_bin(double value, double *bins, int resolution, const double L) { // L is the box size
+    // float index
+    float iflt = (value - bins[0]) / (bins[resolution] - bins[0]) * resolution;
+
+    // round down to the nearest integer
+    int index = (int)floorf(iflt);
+
+    // check if the value is within the range
+    if (index >= 0 && index < resolution) {
+        return index;
+    }
+    else {
+        return -1;
+    }
+    endrun(0, "Error in find_bin\n");
+}
+
+void grid3d_nfw(const struct particle_data * Parts, int num_particles, double **binning, GridDimensions dims, double *density) {
+    
+    double position[3];
+    // Process each particle
+    for (int p = 0; p < num_particles; p++) {
+
+        // remove offset
+        for(int d = 0; d < 3; d ++) {
+            position[d] = Parts[p].Pos[d] - PartManager->CurrentParticleOffset[d];
+            while(position[d] > PartManager->BoxSize) position[d] -= PartManager->BoxSize;
+            while(position[d] <= 0) position[d] += PartManager->BoxSize;
+        }
+        int ix = find_bin(position[0], binning[0], dims.nx, PartManager->BoxSize);
+        int iy = find_bin(position[1], binning[1], dims.ny, PartManager->BoxSize);
+        int iz = find_bin(position[2], binning[2], dims.nz, PartManager->BoxSize);
+
+        // continue if the particle is outside the grid
+        if (ix == -1 || iy == -1 || iz == -1) {
+            continue;
+        }
+        // Increment the density in the appropriate bin
+        ACCESS_3D(density, ix, iy, iz, dims.ny, dims.nz)++;
+    }
+
+}
+
+void projectDensity(double *density, double *density_projected, GridDimensions dims, int normal) {
```

Comment:
I think the best way to keep the code general but still fast is to keep grid3d_nfw as is, but make this function do a reduce in place along the z directrion. Thus it would be a no-op for thickness = 1. 

So something like:
for (k = 1 ; k < DimNorm; k++) //Note that this does nothing if there is only one z-bin!
for(i)
for (j)
 ACCESS_3D(density_projected, i, j, 0, Dim1) += ACCESS_3D(density, i, k, j, dims.ny, dims.nz);

If you arrange the memory right you can use realloc() afterwards to shrink the memory (although in this case there is no need for that)

Reply:
yes, done!